libraries/Adapters/Microsoft.Bot.Builder.Adapters.Twilio/TwilioAdapter.cs

```diff
@@ -0,0 +1,243 @@
+
+            if (!twilioSignature.Equals(string.Empty) && requestValidator.Validate(validationUrl, body, twilioSignature))
+            {
+                var jsonBody = JsonConvert.SerializeObject(body);
+
+                TwilioEvent twilioEvent = JsonConvert.DeserializeObject<TwilioEvent>(jsonBody);
+
+                var activity = new Activity()
+                {
+                    Id = twilioEvent.MessageSid,
+                    Timestamp = new DateTime(),
+                    ChannelId = "twilio-sms",
+                    Conversation = new ConversationAccount()
+                    {
+                        Id = twilioEvent.From,
+                    },
+                    From = new ChannelAccount()
+                    {
+                        Id = twilioEvent.From,
+                    },
+                    Recipient = new ChannelAccount()
+                    {
+                        Id = twilioEvent.To,
+                    },
+                    Text = (twilioEvent as dynamic).Body,
+                    ChannelData = twilioEvent,
+                    Type = ActivityTypes.Message,
+                };
+
+                // Detect attachments
+                if (Convert.ToInt32(twilioEvent.NumMedia) > 0)
+                {
+                    // specify a different event type for Botkit
+                    (activity.ChannelData as dynamic).botkitEventType = "picture_message";
+                }
+
+                // create a conversation reference
+                using (var context = new TurnContext(this, activity))
+                {
+                    await RunPipelineAsync(context, bot.OnTurnAsync, default);
+
+                    response.StatusCode = 200;
+                    response.ContentType = "text/plain";
+                    var text = (context.TurnState["httpBody"] != null) ? context.TurnState["httpBody"].ToString() : string.Empty;
+                    await response.WriteAsync(text);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Standard BotBuilder adapter method to update a previous message with new content.
+        /// </summary>
+        /// <param name="turnContext">A TurnContext representing the current incoming message and environment.</param>
+        /// <param name="activity">The updated activity in the form '{id: `id of activity to update`, ...}'.</param>
+        /// <param name="cancellationToken">A cancellation token for the task.</param>
+        /// <returns>A resource response with the Id of the updated activity.</returns>
+        public override async Task<ResourceResponse> UpdateActivityAsync(ITurnContext turnContext, Activity activity, CancellationToken cancellationToken)
+        {
+            // Twilio adapter does not support updateActivity.
+            return await Task.FromException<ResourceResponse>(new NotSupportedException("Twilio SMS does not support updating activities."));
```

Comment:
This one is new to me, I never used Task.FromException.
What is the difference between using this or just doing 

```
throw new NotSupportedException("Twilio SMS does not support updating activities.")
```

If we use Task.FromException, let's make sure we add ConfigureAwait(false) at the end of it.

Reply:
Feedback applied. Thanks!